### PR TITLE
Add test on iOS simulator

### PR DESCRIPTION
**Pull request by sunmou99**
Script that running game-loop test on simulator: scripts/gha/test_simulator.py
A UI Test App that enable game-loop test: scripts/gha/integration_testing/gameloop

Workflow passed (admob,analytics):
https://github.com/firebase/firebase-cpp-sdk/runs/2241571055?check_suite_focus=true#step:33:28

Many test failed on simulator (due to keychain issue):
https://github.com/firebase/firebase-cpp-sdk/runs/2241582501?check_suite_focus=true#step:33:31

### CODE REVIEW

**Review by jonsimantov — COMMENTED** [review]
What is this file? scripts/gha/integration_testing/gameloop.zip

**Comment by sunmou99** [comment]
> What is this file? scripts/gha/integration_testing/gameloop.zip

I built the app & XCUITest, and compressed the test files into a .zip file. 
Thus we could skip the build gameloop (app & test) step and save some time.

**Comment by jonsimantov** [comment]
Rather than letting our tests go red, could you add a flag that turns off the simulator tests, so that we can file a ticket to re-enable and fix them?

File: scripts/gha/integration_testing/gameloop/gameloopUITests/gameloopUITests.swift [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+//  limitations under the License.
+//
+//  gameloopUITests.swift
+//  gameloopUITests
+//
+
+import XCTest
+
+/// Constants used across the loop launcher and UI test.
+public struct Constants {
```
Comment (vimanyu) [inline scripts/gha/integration_testing/gameloop/gameloopUITests/gameloopUITests.swift:23]: Don't know much about swift but noticed the same constants are also defined in "Constants.swift". Could we reuse?

File: scripts/gha/test_simulator.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+  """Run the gamelop test."""
+  args = ["xcodebuild", "test-without-building", 
+    "-xctestrun", gameloop_app, 
+    "-destination", "id=%s" % device_id]
+  logging.info("Running game-loop test: %s", " ".join(args))
+  result = subprocess.run(args=args, capture_output=True, text=True, check=False)
+
+  if not result.stdout:
+    logging.info("No xctest result")
+    return None
```
Comment (vimanyu) [inline scripts/gha/test_simulator.py:234]: nit: None is redundant. could just say "return".

**Review by vimanyu — COMMENTED** [review]
Minor comments on python code.

**Comment by jonsimantov** [comment]
> > What is this file? scripts/gha/integration_testing/gameloop.zip
> 
> I built the app & XCUITest, and compressed the test files into a .zip file.
> Thus we could skip the build gameloop (app & test) step and save some time.

I don't really like including a prebuilt app in our repo - is this zip file used by the script? Or does the script build the app?

File: scripts/gha/test_simulator.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        return os.path.join(file_dir, file_name)
+  
+  return None
+
+
+def _boot_simulator(ios_device):
+  """Create a simulator locally. Will wait until this simulator botted."""
+  device_info = ios_device.split("-")
+  device_name = device_info[0]
+  device_os = device_info[1]
```
[the remark this answers is not in the document]
Comment (sunmou99) [reply scripts/gha/test_simulator.py:150]: > > > What is this file? scripts/gha/integration_testing/gameloop.zip
> > 
> > 
> > I built the app & XCUITest, and compressed the test files into a .zip file.
> > Thus we could skip the build gameloop (app & test) step and save some time.
> 
> I don't really like including a prebuilt app in our repo - is this zip file used by the script? Or does the script build the app?

The script uses this zip file

**Comment by sunmou99** [comment]
> > > What is this file? scripts/gha/integration_testing/gameloop.zip
> > 
> > 
> > I built the app & XCUITest, and compressed the test files into a .zip file.
> > Thus we could skip the build gameloop (app & test) step and save some time.
> 
> I don't really like including a prebuilt app in our repo - is this zip file used by the script? Or does the script build the app?

The script uses this zip file

File: scripts/gha/test_simulator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunmou99) [reply scripts/gha/test_simulator.py:150]: I checked, no device name (both Android & iOS) contains "-".
But I noticed that setting device info could be a problem.
My next plan is to expend the test device matrix. And I will figure out how to deal with it then.

File: scripts/gha/test_simulator.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+  if len(argv) > 1:
+    raise app.UsageError("Too many command-line arguments.")
+
+  current_dir = pathlib.Path(__file__).parent.absolute()
+  testapp_dir = os.path.abspath(os.path.expanduser(FLAGS.testapp_dir))
+  gameloop_zip = os.path.join(current_dir, FLAGS.gameloop_zip)
+  ios_device = FLAGS.ios_device
+
+  config_path = os.path.join(current_dir, "integration_testing/build_testapps.json")
+  with open(config_path, "r") as config:
```
[the remark this answers is not in the document]
Comment (sunmou99) [reply scripts/gha/test_simulator.py:85]: Historically, testing scripts name variable this way.

**Comment by sunmou99** [comment]
> Rather than letting our tests go red, could you add a flag that turns off the simulator tests, so that we can file a ticket to re-enable and fix them?
> 
> (I recommend an env setting since we are out of input parameters.)

Add an env, and modified the build script to support it.
Also, test device takes 4 inputs. I think is better to address this issue in the following expending test device matrix ticket.

File: scripts/gha/test_simulator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vimanyu) [reply scripts/gha/test_simulator.py]: Sorry, I meant device_info could be missing a "-". and the token split will fail in that case.

[the remark this answers is not in the document]
Comment (sunmou99) [reply scripts/gha/test_simulator.py]: Got it. Currently, we only use the default value, thus it won't be a problem. 
But I will keep this in mind when expending the test device matrix.

**Review by jonsimantov — APPROVED** [review]

**Pull request merged**